delivery-service/ds/src/main.rs

```
@@ -0,0 +1,305 @@
+
+    let mut data = data.lock().unwrap();
+    let client_name = info.client_name.clone();
+    let old = data.clients.insert(info.id.clone(), info);
+    if old.is_some() {
+        return actix_web::HttpResponse::Conflict().finish();
+    }
+
+    actix_web::HttpResponse::Ok().body(format!("Welcome {}!\n", client_name))
+}
+
+/// Returns a list of clients with their names and IDs.
+#[get("/clients/list")]
+async fn list_clients(_req: HttpRequest, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    log::debug!("Listing clients");
+    let data = data.lock().unwrap();
+
+    // XXX: we could encode while iterating to be less wasteful.
+    let clients: Vec<ClientInfo> = data.deref().clients.values().map(|c| c.clone()).collect();
+    let mut out_bytes = Vec::new();
+    if clients.encode(&mut out_bytes).is_err() {
+        return actix_web::HttpResponse::InternalServerError().finish();
+    };
+    actix_web::HttpResponse::Ok().body(Body::from_slice(&out_bytes))
+}
+
+/// Resets the server state.
+#[get("/reset")]
+async fn reset(_req: HttpRequest, data: web::Data<Mutex<DsData>>) -> impl Responder {
+    log::debug!("Resetting server");
+    let mut data = data.lock().unwrap();
+    let data = data.deref_mut();
+    data.clients.clear();
+    data.groups.clear();
+    actix_web::HttpResponse::Ok().finish()
+}
+
+/// Get the list of key packages for a given client `{id}`.
+/// This returns a serialised vector of `ClientKeyPackages` (see the `ds-lib` for
+/// details).
+#[get("/clients/key_packages/{id}")]
+async fn get_key_packages(
+    web::Path(id): web::Path<String>,
+    data: web::Data<Mutex<DsData>>,
+) -> impl Responder {
+    let data = data.lock().unwrap();
+
+    let id = match base64::decode_config(id, base64::URL_SAFE) {
+        Ok(v) => v,
+        Err(_) => return actix_web::HttpResponse::BadRequest().finish(),
+    };
+    log::debug!("Getting key packages for {:?}", id);
+
+    let client = data.clients.get(&id).unwrap();
```

This should probably return an error if there is no match for that id.